static/js/filter.js

```diff
@@ -36,43 +37,74 @@
-        const matchListAliasLabels = filterAliasLabels.filter(s => s.includes(query));
+        const matchListAliasLabels = searchIndexWithLabels.filter(item => item.label.includes(query));
 
         // Set the current number of "found" results
         if ( (matchListEmailAddresses.length + matchListAliasLabels.length) <= aliases.length ) {
             filterLabelVisibleCases.textContent = matchListEmailAddresses.length + matchListAliasLabels.length;
         }
 
+        // Show email addresses that match the search query
         for (const alias of matchListEmailAddresses) {
             let index = filterEmailAddresses.indexOf(alias);
             if (index >= 0) {
                 aliasCollection[index].style.display = "block";
             }
         }
 
-        // TODO: Map the entire list of aliases better so we dont have two seperate DOM collection arrays to loop through.
-        for (const alias of matchListAliasLabels) {
-            let index = filterAliasLabels.indexOf(alias);
-            if (index >= 0) {
-                aliasesWithLabelsCollection[index].style.display = "block";
-            }
+        // Show aliases with labels that match the search query
+        for (const result of matchListAliasLabels) {        
+            result.alias.style.display = "block";
         }
 
 	}
 
     // This function catches any label updates from the user and updates the search array to display them.
     function updateAliasLabel(event){
+        const alias = event.target.closest(".js-alias");
         const prevLabel = event.target.dataset.label;
-        const newLabel = event.target.value.toString().toLowerCase();
-        const labelArrayIndex = filterAliasLabels.indexOf(prevLabel);
+        const prevLabelLowercased = prevLabel.toString().toLowerCase();
+        const newLabel = event.target.value;
+        const newLabelLowercased = newLabel.toString().toLowerCase();
+        const AliasWithLabelsArrayIndex = aliasesWithLabelsCollection.indexOf(alias);
+        const labelArrayIndex = filterAliasLabels.indexOf(prevLabelLowercased);
+
+
+        // Case: User did not enter any label, nor was one previously set
+        if (prevLabel === "" && newLabel === "") {
+            return;
+        }
 
+        // Case: User deleted/cleared an existing label
+        if (newLabel === "") {
+            filterAliasLabels.splice(labelArrayIndex, 1);
+            aliasesWithLabelsCollection.splice(AliasWithLabelsArrayIndex, 1);
+            return;
+        }
+
+        // Case: User updated an existing label to a new string
         if (labelArrayIndex > -1) {
-            filterAliasLabels.splice(labelArrayIndex, 1, newLabel);
+            filterAliasLabels.splice(labelArrayIndex, 1, newLabelLowercased);
+            return;
         }
+
+        // Case: User created a label for an alias
+        aliasesWithLabelsCollection.push(alias);
+        filterAliasLabels.push(newLabelLowercased);
+
```

Comment:
This function got a major update as well – it now can account for all the different cases where a user interacts with the label input form. _(Creating a new label, changing an existing one, or removing an existing one)_